src/item_location.cpp

```diff
@@ -31,7 +33,12 @@ class item_location::impl
         virtual void remove_item() = 0;
 
     protected:
-        item *what = nullptr;
+        impl( item_uid &&id ) : id( std::move( id ) ) {}
+
+        virtual item *target() const = 0;
+
+        mutable item *what = nullptr;
```

Comment:
`mutable` is required because `what` is calculated from `id` the first time `target()` runs. This caching both improves the efficiency and simplicity of the class. Mutability is a promise that the consumers of the class can't detect this internal change in state.

The class can't set `what` at item construction without enforcing a specific load order on JSON. I can drop `mutable` but every call to `target()` then becomes a lookup which would be inefficient